src/lib/consensus/proof_of_stake_fuzzer.ml

```diff
@@ -0,0 +1,518 @@
+          Tick.prove
+            (Tick.Keypair.pk Keys.Step.keys)
+            (Keys.Step.input ()) prover_state main next_state_top_hash
+        in
+        { Blockchain.state= next_state
+        ; proof= wrap next_state_top_hash prev_proof } )
+  in
+  Or_error.iter_error res ~f:(fun e ->
+      Logger.error logger ~module_:__MODULE__ ~location:__LOC__
+        ~metadata:[("error", `String (Error.to_string_hum e))]
+        "Prover threw an error while extending block: $error" ) ;
+  res
+
+[%%elif
+proof_level = "check"]
+
+let genesis_protocol_state =
+  Genesis_protocol_state.t ~genesis_ledger:Test_genesis_ledger.t
+
+let prove_blockchain ~logger (module Keys : Keys_lib.Keys.S)
+    (chain : Blockchain.t) (next_state : Protocol_state.Value.t)
+    (block : Snark_transition.value) state_for_handler pending_coinbase =
+  let next_state_top_hash = Keys.Step.instance_hash next_state in
+  let prover_state =
+    { Keys.Step.Prover_state.prev_proof= chain.proof
+    ; wrap_vk= Tock.Keypair.vk Keys.Wrap.keys
+    ; prev_state= chain.state
+    ; expected_next_state= Some next_state
+    ; update= block
+    ; genesis_state_hash= With_hash.hash genesis_protocol_state }
+  in
+  let main x =
+    Tick.handle (Keys.Step.main ~logger x)
+      (Consensus.Data.Prover_state.handler state_for_handler ~pending_coinbase)
+  in
+  let res =
+    Or_error.map
+      (Tick.check
+         (main @@ Tick.Field.Var.constant next_state_top_hash)
+         prover_state)
+      ~f:(fun () ->
+        {Blockchain.state= next_state; proof= Precomputed_values.base_proof} )
+  in
+  Or_error.iter_error res ~f:(fun e ->
+      Logger.error logger ~module_:__MODULE__ ~location:__LOC__
+        ~metadata:[("error", `String (Error.to_string_hum e))]
+        "Prover threw an error while extending block: $error" ) ;
+  res
+
+[%%elif
+proof_level = "none"]
+
+let prove_blockchain ~logger:_ _ _ _ _ _ _ =
+  failwith "cannot run fuzzer with proof_level = \"none\""
+
+[%%else]
+
+[%%show
+proof_level]
+
+[%%error
+"invalid proof_level"]
+
+[%%endif]
+
+(* TODO: update stakers' relative local_states *)
+let propose_block_onto_chain ~logger ~keys
+    (previous_transition, previous_staged_ledger) (proposer_pk, block_data) =
+  let open Deferred.Let_syntax in
+  let proposal_slot = Block_data.global_slot block_data in
+  let proposal_time = Global_slot.start_time proposal_slot in
+  let previous_protocol_state =
+    External_transition.protocol_state previous_transition
+  in
+  let previous_protocol_state_body_hash =
+    Protocol_state.body previous_protocol_state |> Protocol_state.Body.hash
+  in
+  let previous_ledger_hash =
+    Protocol_state.blockchain_state previous_protocol_state
+    |> Blockchain_state.snarked_ledger_hash
+  in
+  let previous_protocol_state_proof =
+    External_transition.protocol_state_proof previous_transition
+  in
+  (* TODO: insert random txns into the pool every block *)
+  let transactions_by_fee = Sequence.empty in
```

Comment:
Note: Need transactions to test different states of the pending coinbase because of how the scan state gets updated, especially when transactions are added to two different trees and when coinbase is split. (There is, however a unit test in staged-ledger that tests this)

Reply:
Yes, that is necessary. I just haven't implemented it yet. Real transactions, realistic vrf distributions, and local state tracking are all TODO items for the blockchain fuzzer before it is very useful.